Add strict code style

Timonwa opened the pull request:
@jennydaman, I have added the rules as you did in PR https://github.com/FNNDSC/ChRIS_store_ui/pull/229.
Fixes #698 
- Please confirm if I missed any rule or if I need to remove any.
- Also, there was already a `.eslintrc.js` file, so I didn't create a `.eslintrc` file.

- I ran the `lint script`, which gave many errors and warnings.
![lint errors](https://user-images.githubusercontent.com/63044364/195763659-1b055e8d-6e3f-45a3-bfbb-1f92e8e1a9ec.png)

- Then I ran `lint:fix` to fix the ones that can be fixed automatically.
![lint errors2](https://user-images.githubusercontent.com/63044364/195763660-eaf539d8-3918-4960-a333-2f902ec936ae.png)

The app is still working fine 😌

- I also ran `npx prettier --write .`  so the printer rule too can be enforced.

But as I write the pull request, I think I should have asked if I was just to add the rules/config and confirm if the rules are okay before implementing them. Please what should I do about this? Should I create a separate pull request for adding the config and another pull request for implementing the config rules, or should i leave it as i have currently done?

**Side note**: For the errors that were not fixed automatically, can I attempt to fix the errors that I can fix manually?

Timonwa commented:
I'm not sure what this block of code does. Should I include it in the .eslintrc file?
```
globals: {
    Atomics: "readonly",
    SharedArrayBuffer: "readonly",
  },
```

jennydaman commented:
@PintoGideon is the primary developer of this codebase so it comes down to his opinion. The biggest questions are: quotes, single or double? semicolons: yes or no?

JulianaSau commented:
Hi @Timonwa, our issues are related and I've been tasked with automating the formatting as per issue #685. We can work together on this so that our changes integrate without issues

Timonwa commented:
Ok. @JulianaSau. I have added the lint and prettier rules already. I am going to resolve update this branch and resolve any pending conflicts. So that you can pull my branch and we can both work on it.

JulianaSau commented:
That would be great. I'll pull your changes once the conflicts are resolved.

We'll see how we can share the work. We can get some pointers from the linked prs where it was done before

Timonwa commented:
Hi @JulianaSau, I have updated the branch and resolved the conflicts. For the `pre commit` and pre `push` hooks, how do you want to go about that? So we can add that before we start fixing the errors.

JulianaSau commented:
For the hooks it isn't going to cause much of an issue since it will work with most of the code that you've written. So I can set up `husky` with `lint-staged`  so that `eslint` and `prettier` only target staged changes and commits that are ready to be pushed to the remote repository.

Then from there we can resolve the issues that are still there in the codebase

Timonwa commented:
ok. you can go ahead and add it then. and let me know when it is done.

JulianaSau commented:
Just an update. I've been having some issues with my npm cache so setting up the project has been a bit challenging. I'll commit my changes tomorrow after I've fixed the issue

Timonwa commented:
No problem @JulianaSau .

jennydaman commented:
BTW @PintoGideon is on vacation this week, so I'll just make some decisions for him.

jennydaman commented on src/.prettierrc:
```diff
@@ -0,0 +1,5 @@
+{
```
Correct me if I'm wrong, but the `.prettierrc` file should exist in the project root, not under `src/`?

https://prettier.io/docs/en/configuration.html

JulianaSau replied:
Yes. It should be in the root.

Timonwa replied:
It should be. that was a mistake.

jennydaman commented:
Let me know if you have any ideas how best to merge this PR. I am thinking it might be safer if we could `git reset` the linting done to `src/`, and then give a heads up to everyone in the other open PRs that we want to merge this one first.

Timonwa commented:
Can you explain further how you want us to use the `git reset`? Which branch and commit would we be reverting to?

JulianaSau commented:
Hi @Timonwa. I've made the changes but you'll have to add me as a collaborator to your repo so that I can push the changes to your fork and to this PR

JulianaSau commented:
> I have added you as a collaborator. @JulianaSau

I've pushed the changes😁. 
I added scripts to `package.json` incase the formatting and linting checks needs to be run with github actions or globally for the whole project
Installed `tsc-files` to do a type check on staged typescript files
Created a `lint-staged.config.js` file that contains actions that can be run for staged files(formatting, linting or unit tests`.
added some files to the `prettierignore` file


So the linting situation is the next order of business. Doing a `git reset` might undo most of the changes that we made up until this point😅

JulianaSau commented:
> I still don't fully get the git reset. We would be resetting every other persons branch to a particular commit based on this repo? Then should I still continue resolving the merge conflicts for this branch?

Not exactly. Since you made the lint changes on this fork, you wouldn't technically be resetting the remote repository. The `git reset` he is suggesting would take this branch back to the point in time before you ran the `lint --fix`  command on the `src` folder([this](https://github.com/FNNDSC/ChRIS_ui/pull/700/commits/fe15370bbb414105fe44c38301e625637608334b)). Which would mean that any commit we've made after that will be discarded. Since fixing the errors could create mass merge conflict for people currently working on the repo, I guess the reset would make sense. But its kind of a temporary solution.

Timonwa commented:
Ok. So we revert back to b4 I made the lint fix. Add the rules, merge to
master, tell ppl to pause on contributions, do the formatting and stuffs,
then everyone pulls d new code and continue contributing?

Timonwa

On Wed, Oct 19, 2022, 9:25 PM Juliana Sau ***@***.***> wrote:

> I still don't fully get the git reset. We would be resetting every other
> persons branch to a particular commit based on this repo? Then should I
> still continue resolving the merge conflicts for this branch?
>
> Not exactly. Since you made the lint changes on this fork, you wouldn't
> technically be resetting the remote repository. The git reset he is
> suggesting would take this branch back to the point in time before you ran
> the lint --fix command on the src folder(this
> <https://github.com/FNNDSC/ChRIS_ui/pull/700/commits/fe15370bbb414105fe44c38301e625637608334b>).
> Which would mean that any commit we've made after that will be discarded.
> Since fixing the errors could create mass merge conflict for people
> currently working on the repo, I guess the reset would make sense. But its
> kind of a temporary solution.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/FNNDSC/ChRIS_ui/pull/700#issuecomment-1284532992>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APA7WDBIZ2SNDYJOWMXYVOLWEBKMHANCNFSM6AAAAAARE4RJIE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

JulianaSau commented:
Exactly. Then the linting errors that are in the repo can be dealt with after we've enforced the new code style

Timonwa commented:
ok. let us hear @jennydaman's opinion before we do the git reset.

nanfuka commented:
> Ok. we can do the git reset then and add the rules again. Then allow @jennydaman to do the merge and run the auto-fix before we fix the rest manually. Does that sound good?

i think that is ok. my question is did u consider console logs detectable by prettier

JulianaSau commented:
> > Ok. we can do the git reset then and add the rules again. Then allow @jennydaman to do the merge and run the auto-fix before we fix the rest manually. Does that sound good?
> 
> i think that is ok. my question is did u consider console logs detectable by prettier

I don't think that rule has been set. That rule could be added to the eslint file

JulianaSau commented:
> Ok. we can do the git reset then and add the rules again. Then allow @jennydaman to do the merge and run the auto-fix before we fix the rest manually. Does that sound good?

Yes. That's okay

Timonwa commented:
hi @jennydaman, and @JulianaSau, I did the git reset. But it brought out more problems. Because I have been periodically updating this branch and resolving merged conflicts, it was hard to know which code changes to keep and which to leave. so I just created a new branch `NewAddStrictStyleCode`, added only the style rules and created a new pull request. #719

I added the new code and the pre-commit rule ran, but I noticed that it didn't run the Prettier along with it (if it ran now though, that would format all the files again and cause issues we were trying to avoid. but after we merge to master, it is supposed to auto-run too on pre-commit. That's how it should work right?

Then also when I tried to push, I couldn't and couldn't figure it out why it wasn't going. So I commented out the code in the pre-push file, so I would be able to push the new branch. @JulianaSau can you take a look at the pre-commit and push codes again, and also check that I copied your codes correctly?

This is the error below:
![ice_screenshot_20221021-112539](https://user-images.githubusercontent.com/63044364/197174750-8c399db3-0d37-42b2-9677-14e09c29c002.png)

JulianaSau commented:
Hi @Timonwa. Sorry for the late response

I'll fix the error on the pre-push hook today. Idk why it's having an error because it passed the last time I was pushing code. We don't need it for now anyway but I'll adjust it in case there's tests that need to be run before code is committed

Timonwa commented:
Thank you.

JulianaSau commented:
The `prettier` definitely ran. The `pre-commit` rule only runs on staged files, not the whole root folder. Since the script didn't detect any js or ts files, it skipped the formatting rule.

 I added scripts for checking and fixing linting and formatting errors on the whole project in the `package.json` for the purpose of GitHub workflows.

Timonwa commented:
That's cool. Then. Thank you.

JulianaSau commented:
> That's cool. Then. Thank you. Please don't forget to work on the new branch.

Okay. I'll inform you once I've made the changes

Timonwa commented:
I am closing this pull request, so it doesn't mistakenly get merged since we have opened a new one for this issue.
